### ISSUE TITLE
Merging in missing functionality from Cadair/jupyter_environment_kernels

### ISSUE THREAD

**Issue by jankatins** [Anaconda-Platform/nb_conda_kernels#42]
There are currently two implementaiton of the functionality added with this package: this here and https://github.com/Cadair/jupyter_environment_kernels/tree/master/environment_kernels. I see it that this implementation will gain a lot of users because of the integration with anaconda and its installed default packages.

On the other hand, the "other" implementation has some features which I'm missing in this implementation. 
- Proper activating the envs (e.g. getting a proper PATH even when having activate.d scripts)
- Only add kernel if the actual kernel functionality is in the env
- virtualenv support

While I think that the first two will eventually be accepted/ added here, I would like to discuss / ask if you would consider adding virtualenv functionality.

I'm willing to add all three things by porting them from https://github.com/Cadair/jupyter_environment_kernels/

CC: @Cadair

[Just to make sure: I think this package also has a few things better than the other one: tests, an automatic installer, ... :-) ]

**Comment by damianavila** [comment]
I think the hard point is the virtualenv support... one crazy thing coming to my mind is to make virtualenv support and "extension" over nb_conda_kernels... some kind of plugin that can be added to enable virtualenv if the user wants that... but otherwise keep the main package only with conda env support.
In any case, we can also discuss support in the nb_conda_kernels itself, my first reaction would be no... also we should change the name if we agree to do it :stuck_out_tongue:

@bollwyvl thoughts?

**Comment by janschulz** [comment]
If support for "other" envs is added, the actual support for them is probably two more function (one to search for them, one to make the "activate" work)  and the config endpoint (where to look for them) in the end. 

I would really appreciate it if there would only be one project with the combined feature set, so that work is not wasted to implement the features in the other project each time... I also wouldn't mind porting the features of this project over to the other one, but I guess it's better to use this here as this gets more exposure due to anaconda inclusion... :-)

**Comment by damianavila** [comment]
> If support for "other" envs is added, the actual support for them is probably two more function (one to search for them, one to make the "activate" work) and the config endpoint (where to look for them) in the end.

Well... if that is the case, I would probably migrate from "no" into "maybe"... :wink:
One of the things I wondering is which is the impact of a conda-scoped extensions for someone using virtualenvs (and then probably not using conda)? Also, if you bring virtualenv support I guess I can ask you to support it in the long term, am I right? :wink: (the real thing is that I don't want to fight again with virtualenv if I have something better like conda envs, jeje). 

>I would really appreciate it if there would only be one project with the combined feature set, so that work is not wasted to implement the features in the other project each time... 

Yes, I agree.. if possible, we should end up with just one.

**Comment by bollwyvl** [comment]
Sorry, late to the party!

>Proper activating the envs (e.g. getting a proper PATH even when having activate.d scripts)

yeah, this is important! I considered the "generate me a fake shell command" approach, but do wonder if this is a thing that we want as a separate activity/package. In the days of yore, there was a `conda run -n foo -- cmd`, but I've seen some other approaches. Endgame would be to get this in conda itself, as it is such a common activity, and so fraught with peril, as we have seen with the 4.x line.

> Only add kernel if the actual kernel functionality is in the env

Also yeah. I had tried this on a development branch, but found it even slower than looking for binaries. But it is the _right_ way.

> virtualenv support

I was a virtualenv user for many years, loved what it did for me. But Anaconda isn't going to support anything that touches virtualenvs, I'd wager.

Perhaps we should think about how to generalize the capability (potentially into something that doesn't do anything other than the `super` defaults) then extend for multiple implementations.

Today there is `conda`, `virtualenv` and then `pyenv`... but then the raft of things for other langs like `nvm`, `rvm` and who-knows-what-else. 

Add to this the capability of something like [remotekernel](https://github.com/danielballan/remotekernel) or [rk](https://github.com/korniichuk/rk), and we'll quickly end up where someone can't have what they want to get the work done on a project that needs multiple approaches without a lot of environment switching... which is exactly what we're trying to avoid.

It has been mentioned several times that at least in the near future, environment management isn't going to show up in Jupyter itself, so it is indeed up to us. And, as it is, one can only have _one_ `KernelSpecManager`, so we have to get along.

So strawman, based on everyone's suggestions above:

`nb_env_kernels` (or whatever) would be the high-level package, and actually be a serverextension and nbextension, providing the UI for searching through kernels, refreshing the list of kernels, etc. (which is why all the caching noise now)

It would expose and read (at startup) an `entry_point` (i suppose we could use the config system, but I think this might be more appropriate here) that would use the key to build up the kernel string and any other things it might need.

All the default kernel names would get grandfathered in without fanciness, as we've learned we can't go messing with those for various reasons. It would handle all the caching, spawning, etc. so that it is done in a consistent way (and all the handles get closed properly on shutdown, etc. which is why i have shied away from it thus far).

Then, `nb_conda_kernels` (or whatever) would be a small `entry_point` that depended on `nb_env_kernels`... probably just one class, with some API that we define. Aside from `discover`, we'd probably want to add some stuff like an icon to pretty up and distinguish between a conda env and a virtualenv or whatever. If we can avoid them being nbextensions/serverextensions, that would be great. Anyhooo... nb_conda_kernels (and by extension, nb_env_kernels) would end up in the Anaconda distribution, and anybody would be free to build them their own selves elsewhere.

Thoughts?

**Comment by janschulz** [comment]
If an plugin architecture is what gets this accepted into this project, I'm all for it...

**Comment by Cadair** [comment]
Hey all, been meaning to get to this for a while...

I still haven't had a detailed look at the code in this repo, so if I misunderstand anything I do apologise. As far as I can tell there is a hunk of overlap between this and the implementation in my repo, mainly, it provides a kernelspecmanager and calls conda to find envs then generates the specs etc. In my repo there is much more configuration options and functionality, I actually started my project to just look for directories, I would point it at the conda env root and it would look inside `/bin` for all the envs, which is why adding virtualenv support was easy * . Also we have the full environment sourcing which @janschulz  added etc. This repo has much more robust tests and other important things.

I love the idea of one package to provide the `KernelSpecManager` and then an entry point to register ways of finding the kernels. This also seems much more like something that would get merged into Jupyter than something conda or virtualenv specific. I would suggest that if we want to go that way we make a new repo (probably a modification or fork of this one) which we aim to be completely platform (conda, virtualenv) independent and then re-write a plugin for that approach for conda and another one for virtualenv (This maybe better taken from my repo as we have the path sourcing stuff etc).

* - Aside, I actually use the directory finding functionality myself because I have Jupyter installed via my linux system python and then it finds the conda envs by looking for the paths as the system python doesn't have conda in the path, so from my perspective this is an important feature!!

**Comment by bollwyvl** [comment]
@Cadair thanks for stopping by!

All good things. Glad you like the basic approach.

Perhaps what we are really pushing for is, _in the name of reproducibility_, provider-agnostic environment awareness across the different components that do _kernel stuff with notebooks_: at present, I think that means:
- `notebook` (which we've now done two ways)
- `nbconvert` (which would need some help)
- `kernel_gateway` (haven't even read the code)
- `nbformat` this name-based thing is for the birds...
  - initially, could be something something simple, URI-based (a la `env://<entry_point>/<env_name>`)
    - could be extended to include packages `env://<entry_point>/<env_name>?<what you call things>=<a thing>,<another thing>`
  - but really, having the option to capture the entire env as _actual_ data would be better:
    - too bad all these package managers have their own, independently Turing-complete specifier patterns...
    ```json
    {"type": "<entry_point>",
     "<what you call things>": ["<a thing> <som version thingy>"], 
      "<where you get things>": ["<some place>"]}
    ```
- probably `nbsomethingelse` that I'm not thinking of right now

BONUS: I think the remote kernel case is very interesting as well... remains to be seen whether we try to lump that under this umbrella, but i feel like if we pushed for a real, non-metadata-wild-west addition to nbformat, we'd want to cleanly add all the "you ain't running in kansas anymore" stuff in one go. But now we're getting into [binder](http://mybinder.org/)/[kapsel](https://github.com/conda/kapsel) territory, but inverting it, as both of those think about a notebook that lives in some box, whereas this says the notebook actually can contain the box.

Perhaps the path is:
- Demonstrate an extensible, user-friendly, pip/conda-packageable KernelSpecManager + serverextension + nbextension for empowering the user to make good, reproducible decisions around how they use kernels
  - _don't try to **install** things in envs, though... whole other ball of wax_
- at least two implementations of `entry_point`y `KernelSpecFinders` (or whatever we call them)
  - _or config'd, but, buh, no official CLI-friendly API for anything but server/nb extensions at present, and really don't want to start this kind of work without `conf.d` support_
- run it through the [incubator](https://github.com/jupyter-incubator)/[JEP](https://github.com/jupyter/enhancement-proposals) process, and perhaps get the behavior of the upstream KernelSpecManager changed, as well as support from nbconvert `--execute`.

**Comment by damianavila** [comment]
I think we have a general agreement about this... I like the steps described by @bollwyvl, and if we can get some help from @janschulz  and @Cadair :wink:, I think we can start doing some work on this direction and have something working relatively quick.
Maybe we need to start with the incubator/JEP proposal as the first steps, thoughts?

**Comment by janschulz** [comment]
The [current implementation of the kernel manager class](https://github.com/Cadair/jupyter_environment_kernels/blob/master/environment_kernels/core.py) is probably much what you have in mind to submit upstream.

* remove the virtualenv/conda related knobs
* add an "extension point" for the additional functions which provide the additional kernel specs (what is now the `ENV_SUPPLYER` variable)
* inline the code in the overwritten methods in the original `KernelSpecManager` to take these additional kernel specs into account
* maybe add the blacklist/whitelist functionality (but this could also go to the conda/virtualenv classes itself)

Then a user could install the new "kernel spec provider" python package and add these function in the config instead of overwriting the `KernelSpecManager` itself.

Still doesn't solve the problem with nbconvert not using it...

**Comment by Cadair** [comment]
I think prototyping up a base `KernelSpecManager` and showing that the plugin system can work would be a good start, then moving to a JEP.

As far as the plugins go, I could see a situation where config would be optional, if the base `SpecManager` auto loaded all plugins and ran with default settings we could make it so most people wouldn't need to do any config.

**Comment by bollwyvl** [comment]
> prototyping up a base KernelSpecManager

Here's a strawman (since i could test it easily) of using entrypoints in the current nb_conda_kernels code base:
https://github.com/bollwyvl/nb_conda_kernels/tree/feature/entry_points

Obviously we'd want to bring over much more of the `jupyter_environment_kernels` stuff...

> make it so most people wouldn't need to do any config

Agreed.

As I had mentioned, as we talk about config, etc. the UI will definitely come up. @damianavila had pointed at the "nested submenu" approach, but this is generally regarded as a bad idea, and has been deprecated in bootstrap... not sure if jupyterlab has brought it over, but let's think a bit outside that, perhaps... here's a rehash of a concept I did earlier, but lost:

![mockup](https://cloud.githubusercontent.com/assets/45380/17557638/f7acd6c6-5ee5-11e6-8674-738f54df6256.png)

Though i'll have to see if we'd have to do too much crazy stuff by putting unexpected things in...

**Comment by janschulz** [comment]
I like what I saw in https://github.com/bollwyvl/nb_conda_kernels/tree/feature/entry_points

So the entrypoint for a kernel spec provider is basically a `provider.find_kernel_specs()` which returns a dict `<name> -> <kernelspec instance>` which then gets "muched" into `<spec-finder-name>-<name> -> <kernelspec instance>` 

So what's missing:

* Make it a PR on the original kernel manager class
* caching: IMO it should be done in the manager and done with threads so that there is no performance penalty when the stuff is expired
* nameing: kernels from different entry point contributor should neverthless get similar names.
* discuss with upstream how to display it in the UI 

My preference would be that the manager would do the naming and would reset the kernel spec name to `{provider}-{language}-{name}` and the display in a similar way (I would prefer a config entry, so that I can name my env kernels with a `~` in front so that they are sorted last). Unfortunately this "renaming" makes only sense if e.g. in the conda case, the r and python finder would be two different entry point contributor as otherwise the names would overlap (if one env would supply two kernels).

**Comment by janschulz** [comment]
In https://github.com/jupyter/nbformat/issues/81 upstream talks about some one pain point of environment kernel managers when used with nbconvert. It might be nice to chime in there (and get the proposal going again :-) ).

**Comment by damianavila** [comment]
Thanks the ping! I was aware but I did not have time to chime in yet.

>It might be nice to chime in there (and get this proposal going again :-) ).

I agree! :wink:

**Comment by mcg1969** [comment]
@jankatins, it looks like there is a bit of duplicated effort between your work and ours (as well as ample reason for yours to continue, e.g., `virtualenv` support). I do have a fix for some of the key issues in #80 that I'm hoping to see merged in.

I also took note of your work to update your package to the new Jupyter 6.0 format. I'm inclined to think that Anaconda—or more precisely, the maintainers of `conda`, no matter what the affiliation—ought to take a strong interest in maintaining conda environment support for jupyter, hence I'm inclined to favor keeping `nb_conda_kernels` around. But I could be swayed...

What I do know is that even if nb_conda_kernels _were_ to be deprecated, it would be important to get these fixes in (in my view) beforehand.

**Comment by mcg1969** [comment]
I've merged https://github.com/Anaconda-Platform/nb_conda_kernels/pull/80 now, so environments are constructed directly from their native kernelspecs, and full activation is performed.
